Support non-stop bookie data migration and bookie offline

lordcheng10 opened the pull request:
### Motivation
bookie offline steps:

1. Log on to the bookie node, check if there are underreplicated ledgers.If there are, the decommission command will force them to be replicated: bin/bookkeeper shell listunderreplicated
2. Stop the bookie : bin/bookkeeper-daemon.sh stop bookie
3. Run the decommission command. If you have logged onto the node you wish to decommission, you don't need to provide -bookieid If you are running the decommission command for target bookie node from another bookie node you should mention the target bookie id in the arguments for -bookieid : bin/bookkeeper shell decommissionbookie or $ bin/bookkeeper shell decommissionbookie -bookieid <target bookieid>
4. Validate that there are no ledgers on decommissioned bookie $ bin/bookkeeper shell listledgers -bookieid <target bookieid>

**For the current bookie offline solution, need to stop the bookie first，execute the decommission command and wait for the ledger migration on the bookie to complete.**

**It can be seen that it is very time-consuming to offline a bookie node. When we need to offline a lot of bookie nodes, the time-consuming of this solution will not be acceptable**

**Therefore, we need a solution that can migrate data without stopping bookie, so that bookie nodes can be offlined in batches**


### Changes

1. Implement a command to submit the bookie to be offline and the corresponding ledgers, for example:
```
bin/bookkeeper shell decommissionbookie -offline_bookieids  bookieId1,bookieId2,bookieId3,...bookieIdn
```
   This command will write all ledgers on the offline bookie node to the zookeeper directory, for example:

 ` put ledgers/offline_replicas/ledgerId  bookId1,bookId2,...bookIdn;`

2. Design a **OfflineReplicasWorker** class to perform the actual ledger replication: this class will obtain a ledger from the zookeeper directory ledgers/offline_replicas for replication. Before copying the data, it will first **filter out all the fragments containing the offline bookieId under the ledger,then copy these fragments**;

In addition, define a concept of Replicas: the replica data of a ledger on one or more bookies.

StevenLuMT commented:
There's some checkstyle error,please update the code like this @lordcheng10 

<img width="2175" alt="image" src="https://user-images.githubusercontent.com/42990025/189253818-32cc3430-96f6-473f-a02c-e9e2121e4a42.png">

lordcheng10 commented:
> There's some checkstyle error,please update the code like this @lordcheng10
> 
> <img alt="image" width="2000" src="https://user-images.githubusercontent.com/42990025/189253818-32cc3430-96f6-473f-a02c-e9e2121e4a42.png">

OK, I will fix!

lordcheng10 commented:
@StevenLuMT fixed,PTAL,thanks!

lordcheng10 commented:
@hangc0276 @eolivelli @zymap PTAL,thanks!

lordcheng10 commented:
rerun failure checks

lordcheng10 commented:
### Motivation
**Currently bookkeeper does not support data migration, only data recovery.**

We have a scenario where a large number of bookie nodes are offline, and we find that the bookie's offline solution is very time-consuming.

Bookie offline steps are as follows:
1. Set the bookie node to be offline to readOnly;
2. Wait for the Pulsar data on the Bookie node to be offline to expire and delete;
3. When most of the data on these offline nodes is expired and cleaned up, there will still be some data that cannot be expired and deleted;
4. Stop a bookie, and then use the decommission command to migrate the data that has not been expired and cleaned up to the new node:
`bin/bookkeeper shell decommissionbookie -bookieid xx`
5. When the data on one bookie node is migrated, continue to the next bookie node;

Step 4 is very time-consuming. We found that waiting for a bookie data migration to complete, it takes about 1 hour, and we have 125 bookie nodes to be offline.

In addition, step 2 is also very time-consuming, depending on the pulsar retain time, usually more than ten hours.

### Proposal
To solve this problem, we developed a data migration tool.
After having this tool, our offline steps are as follows:
1. Execute the data migration command:
`bin/bookkeeper shell replicasMigration --bookieIds bookie1,bookie2 --ledgerIds ALL --readOnly true`
2. When the data migration is completed, stop all bookie nodes to be offline;

In addition, this command can also migrate the replica data on some bookie nodes to other nodes, for example:

`bin/bookkeeper shell replicasMigration --bookieIds bookie1,bookie2 --ledgerIds ledger1,ledger2,ledger3 --readOnly false`


### For example
1. Migrate all ledger data on bookie1 and bookie2 to other bookie nodes:
`sh bin/bookkeeper shell replicasMigration -bookieIds bookie1,bookie2 -ledgerIds ALL -readOnly true
`
2. Migrate ledger1 and ledger3 on bookie1 and bookie2 to other bookie nodes:
`sh bin/bookkeeper shell replicasMigration -bookieIds bookie1,bookie2 -ledgerIds ledger1,ledger3 -readOnly false`


### Application scenarios:
**1. The bookie node goes offline:**
As mentioned above, after bookkeeper has this data migration tool, the offline steps of bookie are only two steps, and the time-consuming is greatly reduced:
a.Execute the data migration command:
bin/bookkeeper shell replicasMigration --bookieIds bookie1,bookie2 --ledgerIds ALL --readOnly true
b. When the data migration is completed, stop all bookie nodes to be offline;

**2. Expand the bookie node to improve the reading speed of historical data:**
a. When the client consumes historical data a few days ago, it hopes to increase the reading speed of historical data by expanding the bookie node.
b. When we expand the new bookie node to the cluster, the new node can only receive the read and write of new data, and cannot improve the reading speed of historical data.
c. After the data migration tool of bookkeeper, we can migrate some historical data to the new node, let the new node provide some historical data reading, and improve the reading speed of historical data.

lordcheng10 commented:
@dlg99 @merlimat PTAL,thanks!

lordcheng10 commented:
@Shoothzj PTAL,thanks!

lordcheng10 commented:
@nicoloboschi PTAL,thanks!

lordcheng10 commented:
rerun failure checks

lordcheng10 commented:
@lhotari @merlimat PTAL,thanks!

lordcheng10 commented:
@hangc0276 @eolivelli @zymap PTAL,thanks!

lordcheng10 commented:
@michaeljmarshall PTAL,thanks!

Shoothzj commented:
Hi @lordcheng10 , could you please explain more **how you do the data migration** in the description?

lordcheng10 commented:
> 1. sh bin/bookkeeper shell replicasMigration -bookieIds bookie1,bookie2 -ledgerIds ALL -readOnly true

The data migration steps are as follows:
1. Submit the ledger copy to be offline to zookeeper through the ReplicasMigrationCommand command. The directory organization on zookeeper is as follows:

ledgers/replicasMigration/ledgerId1

And record the bookie node where the ledger1 wants to migrate data on the zookeeper directory, for example:

set ledgers/replicasMigration/ledgerId1 bookie1,bookie3

2. Start the corresponding data replication service ReplicasMigrationWorker in AutoRecoveryMain. The ReplicasMigrationWorker service first obtains a ledger copy migration task, then finds the fragments stored by the ledger on the corresponding bookie node, and finally replicates these fragments through the replicateLedgerFragment method.

3. When a ledger copy migration task is completed, the corresponding ledger directory on zookeeper will be deleted, for example:
delete ledgers/replicasMigration/ledgerId1

lordcheng10 commented:
> 

If we want to migrate all ledger data on some bookies, we need to set these bookies to read only first, otherwise during the migration process, new data may be written to these bookies

lordcheng10 commented:
> > 1. sh bin/bookkeeper shell replicasMigration -bookieIds bookie1,bookie2 -ledgerIds ALL -readOnly true
> 
> @Shoothzj
> 
> The data migration steps are as follows:
> 
> 1. Submit the ledger copy to be offline to zookeeper through the ReplicasMigrationCommand command. The directory organization on zookeeper is as follows:
> 
> > `ledgers/replicasMigration/ledgerId1`
> 
> And record the bookie node where the ledger1 wants to migrate data on the zookeeper directory, for example:
> 
> > `set ledgers/replicasMigration/ledgerId1 bookie1,bookie3`
> 
> 2. Start the corresponding data replication service ReplicasMigrationWorker in AutoRecoveryMain. The ReplicasMigrationWorker service first obtains a ledger copy migration task, then finds the fragments stored by the ledger on the corresponding bookie node, and finally replicates these fragments through the replicateLedgerFragment method.
> 3. When a ledger copy migration task is completed, the corresponding ledger directory on zookeeper will be deleted, for example:
> 
> > `delete ledgers/replicasMigration/ledgerId1`

If we want to migrate all ledger data on some bookies, we need to set these bookies to read only first, otherwise during the migration process, new data may be written to these bookies

lordcheng10 commented:
@eolivelli PTAL,thanks!

Shoothzj commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataBookieDriver.java:
```diff
@@ -88,6 +89,13 @@ default CompletableFuture<Void> disableHealthCheck() {
         return result;
     }
 
+    default CompletableFuture<Void> submitToMigrateReplicas(long ledgerId, String bookieIds)
```
This method needs javadoc

lordcheng10 replied:
fixed

lordcheng10 replied:
@Shoothzj PTAL,thanks!

Shoothzj commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataBookieDriver.java:
```diff
@@ -88,6 +89,13 @@ default CompletableFuture<Void> disableHealthCheck() {
     }
 
+    default CompletableFuture<Void> submitToMigrateReplicas(long ledgerId, String bookieIds)
+            throws UnsupportedEncodingException {
```
why we need this exception? If we use **UTF-8**, I think we can remove this throw

lordcheng10 replied:
@Shoothzj PTAL,thanks!

Shoothzj commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java:
```diff
@@ -106,6 +110,23 @@ interface SessionStateListener {
+        return new ArrayList<>();
+    }
+
+    default void lockMigrationReplicas(String lockPath, String advertisedAddress)
```
method in interface needs javadoc

Shoothzj commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicasMigrationWorker.java:
```diff
@@ -0,0 +1,216 @@
+    public void run() {
+        workerRunning = true;
+        while (workerRunning) {
+            //1. Get a replica ledger to replicate
```
```suggestion
            // 1. Get a replica ledger to replicate
```

lordcheng10 replied:
fixed

lordcheng10 replied:
@Shoothzj PTAL,thanks!

Shoothzj commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicasMigrationWorker.java:
```diff
@@ -0,0 +1,216 @@
+            }
+
+            LOG.info(String.format("Start migrate replicas(%s)!", replicas));
+            //2. Get the fragment containing the migration bookie
```
```suggestion
            // 2. Get the fragment containing the migration bookie
```

lordcheng10 replied:
fixed

Shoothzj commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicasMigrationWorker.java:
```diff
@@ -0,0 +1,216 @@
+            try (LedgerHandle lh = admin.openLedgerNoRecovery(replicas.ledgerId)) {
+                Set<LedgerFragment> fragments = getFragmentsOnMigrationBookies(lh, replicas.bookieIds);
+                if (fragments.size() < 1) {
+                    //3.The replication has been completed, delete the ledgerId directory and release lock
```
```suggestion
                    // 3.The replication has been completed, delete the ledgerId directory and release lock
```

Shoothzj commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicasMigrationWorker.java:
```diff
@@ -0,0 +1,216 @@
+                    LOG.info(String.format("Finish ledgerId %s migration!", replicas.ledgerId));
+                }
+
+                //4. Start replicate
```
```suggestion
                // 4. Start replicate
```

lordcheng10 replied:
fixed

lordcheng10 replied:
@Shoothzj PTAL,thanks!

Shoothzj commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,191 @@
+            LOG.error("Received exception in ReplicasMigrationCommand ", e);
+            return false;
+        } finally {
+            if (admin != null) {
```
admin can't be null here. Am I right?

lordcheng10 replied:
fixed

lordcheng10 replied:
@Shoothzj PTAL,thanks!

Shoothzj commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,191 @@
+        for (BookieId bookieId : bookieIdSet) {
+            String bookieAddress = bookieId.getId().split(":")[0];
+            String httpAddress = bookieAddress + ":" + flags.port;
+            String url = String.format("http://%s/api/v1/bookie/state/readonly", httpAddress);
```
we can use the constant in `HttpRouter.BOOKIE_STATE_READONLY`

lordcheng10 replied:
fixed

lordcheng10 replied:
@Shoothzj PTAL,thanks!

Shoothzj commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,191 @@
+            String httpAddress = bookieAddress + ":" + flags.port;
+            String url = String.format("http://%s/api/v1/bookie/state/readonly", httpAddress);
+            String readOnly = flags.readOnly ? "true" : "false";
+            CloseableHttpResponse response = HttpClientUtils.doPutRequest(url, readOnly, "UTF-8");
```
can we reuse some code in `listbookies` command?

lordcheng10 replied:
@Shoothzj PTAL,thanks!

lordcheng10 replied:
The listbookies command is the zookeeper directory of the list, not the http request sent, so there is no way to reuse it!

lordcheng10 replied:
@Shoothzj PTAL,thanks!

lordcheng10 commented:
rerun failure checks

lordcheng10 commented:
@hangc0276 PTAL,thanks!

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java:
```diff
@@ -106,6 +110,27 @@ interface SessionStateListener {
+
+    default String getOwnerBookiesMigrationReplicas(String ledgerForMigrationReplicasPath)
+            throws InterruptedException, KeeperException, UnsupportedEncodingException {
+        return "";
```
I would throw some UnsupportedOperationException

lordcheng10 replied:
OK, I will fix!

lordcheng10 replied:
fixed! @eolivelli

lordcheng10 replied:
fixed. PTAL,thanks! @eolivelli

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java:
```diff
@@ -106,6 +110,27 @@ interface SessionStateListener {
+    }
+
+    default void deleteZkPath(String migrationLedgerPath) throws InterruptedException, KeeperException {
+    }
```
I would throw some UnsupportedOperationException

lordcheng10 replied:
fixed! @eolivelli

lordcheng10 replied:
fixed. PTAL,thanks! @eolivelli

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,190 @@
+        Set<BookieId> bookieIdSet = parseBookieIds(flags.bookieIdsOfMigratedReplica);
+        for (BookieId bookieId : bookieIdSet) {
+            String bookieAddress = bookieId.getId().split(":")[0];
+            String httpAddress = bookieAddress + ":" + flags.port;
```
if the bookie is up and running it publishes on metadata the http endpoint.

lordcheng10 replied:
@eolivelli 
Do you suggest to get the http port of the corresponding bookie from zookeeper?

But sometimes the bookie node may also be down.
So it may be better to let the user input the http port, if you do not enter the http port, use the default port 8080

lordcheng10 replied:
OK , I will fix.

lordcheng10 replied:
fixed.  PTAL,thanks! @eolivelli

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,190 @@
+
+        Set<BookieId> bookieIdSet = parseBookieIds(flags.bookieIdsOfMigratedReplica);
+        for (BookieId bookieId : bookieIdSet) {
+            String bookieAddress = bookieId.getId().split(":")[0];
```
this doesn't work if your bookieId is not a public hostname.
the bookieId is a opaque string, we have to "resolve" the ID

lordcheng10 replied:
@eolivelli 
So this command is more appropriate to pass in the ip address? 
--bookieIds renamed to --bookies,  for example:

bin/bookkeeper shell replicasMigration --bookies IP1:Port,IP2:Port2 --ledgerIds ALL --readOnly true

hangc0276 replied:
+1, we need to use `resolve` the ID

lordcheng10 replied:
OK , I will fix.

lordcheng10 replied:
fixed. PTAL,thanks! @eolivelli

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/util/HttpClientUtils.java:
```diff
@@ -0,0 +1,47 @@
+ * http client utils.
+ * */
+public class HttpClientUtils {
+    private static final CloseableHttpClient httpClient = HttpClientBuilder.create().build();
```
we should close this object
using a static field is not good.

we can open/close the client per each request, it is not a big deal

lordcheng10 replied:
OK, I will fix!

lordcheng10 replied:
fixed! @eolivelli

lordcheng10 replied:
fixed. PTAL,thanks! @eolivelli

eolivelli reviewed (changes requested):
good work

I have left some comments.

we also need tests, especially tests about recovering a partial migration that was interrupted

lordcheng10 commented:
> we also need tests, especially tests about recovering a partial migration that was interrupted

@eolivelli 

In the process of migrating data, stop the bookie process responsible for the migration task? After stopping the corresponding bookie process, other bookie processes will grab the ledger replication task and continue the replication, so I need to verify whether the ledger migration task can be completed normally? For example, check the final number of copies of the ledger and the distribution of copies on the bookie node after migration?

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java:
```diff
@@ -1951,6 +1953,56 @@ public int runCmd(CommandLine cmdLine) throws Exception {
+                    "Bookies corresponding to the migrated replica");
+            opts.addOption("l", "ledgerIds", true,
+                    "The ledgerIds corresponding to the migrated replica");
+            opts.addOption("r", "readOnly", true,
```
1. required = false?
2. We'd better set the default readOnly flag to `false`. If the admin uses this command to decommission some bookies, which configured placement policy, it may lead to writing new entries failing due to not meeting placement policy.

lordcheng10 replied:
@hanggc0276
The readOnly default value has been set to false here:
<img width="1025" alt="image" src="https://user-images.githubusercontent.com/19296967/190057439-77a31577-67a5-4d15-b5d7-c12bb3c354c7.png">

lordcheng10 replied:
OK, I will  fix!
<img width="1117" alt="image" src="https://user-images.githubusercontent.com/19296967/190059205-2ff05bfb-3a73-4c0d-aa32-d3b80d43d3d3.png">

lordcheng10 replied:
Fixed. 
PTAL,thanks!  @hangc0276

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,190 @@
+
+        @Parameter(names = { "-b", "--bookieIds" }, description = "Bookies corresponding to the migrated replica, "
+                + "eg: bookieId1,bookieId2,bookieId3")
+        private String bookieIdsOfMigratedReplica;
```
Use `List<String>` instead of "," separate String.

lordcheng10 replied:
OK, I will fix!

lordcheng10 replied:
Fixed. 
PTAL,thanks!  @hangc0276

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,190 @@
+        @Parameter(names = { "-l", "--ledgerIds" }, description = "The ledgerIds corresponding to the migrated replica,"
+                + "eg: ledgerId1,ledgerId2,ledgerId3. The default is ALL, which means that all replica data on these "
+                + "bookie nodes will be migrated to other bookie nodes")
+        private String ledgerIdsOfMigratedReplica;
```
Use `List<Long>` instead of `String`

lordcheng10 replied:
I will fix

lordcheng10 replied:
Fixed. 
PTAL,thanks!  @hangc0276

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,190 @@
+
+        @Parameter(names = { "-r", "--readOnly" }, description = "Whether to set the bookie nodes of the migrated data"
+                + " to readOnly, the default is true")
+        private boolean readOnly;
```
Please configure the default value to false.

lordcheng10 replied:
OK, I will fix!

lordcheng10 replied:
Fixed. 
PTAL,thanks!  @hangc0276

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,190 @@
+        private boolean readOnly;
+
+        @Parameter(names = {"-p", "--httpPort"}, description = "http port,default is 8080")
+        private String port;
```
The type is `int`?

lordcheng10 replied:
remove port, resovle ID can get port

lordcheng10 replied:
Fixed. 
PTAL,thanks!  @hangc0276

lordcheng10 commented:
> we also need tests, especially tests about recovering a partial migration that was interrupted

@eolivelli 

How do I need to verify that it is as expected?

We have done some simple verification in the test cluster, by checking the number of replicas and bookie distribution after migration of ledger replicas.

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,190 @@
+            if (responseCode != HttpServer.StatusCode.OK.getValue()) {
+                LOG.error(String.format("Set readonly to %s failed! bookieId:%s, responseCode:%s ",
+                        readOnly, bookieId, responseCode));
+                return false;
```
If we want to decommission 3 bookies, <bookie0, bookie1, bookie2>, and <bookie0, bookie1> set to readOnly mode succeed, and bookie2 set to readOnly failed, do we need to rollback <bookie0, bookie1> to the previous mode?

lordcheng10 replied:
OK , add rollback

lordcheng10 replied:
Fixed. 
PTAL,thanks!  @hangc0276

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,190 @@
+            throws BKException, InterruptedException, IOException {
+        ClientConfiguration adminConf = new ClientConfiguration(conf);
+        BookKeeperAdmin admin = new BookKeeperAdmin(adminConf);
+        try {
```
Suggest `try (BookKeeperAdmin admin = new BookKeeperAdmin(adminConf))`

lordcheng10 replied:
Fixed. 
PTAL,thanks!  @hangc0276

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,190 @@
+            for (Map.Entry<BookieId, Set<Long>> bookieIdSetEntry : bookieLedgerMaps.entrySet()) {
+                Set<Long> ledgersOnBookieId = bookieIdSetEntry.getValue();
+                if (!flags.ledgerIdsOfMigratedReplica.equals(ALL)) {
+                    ledgersOnBookieId.retainAll(ledgerIdsToMigrate);
```
`ledgerIdsToMigrate ` maybe null

lordcheng10 replied:
Fixed. 
PTAL,thanks!  @hangc0276

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,190 @@
+            Map<BookieId, Set<Long>> bookieLedgerMaps = admin.listLedgers(bookieIds);
+            for (Map.Entry<BookieId, Set<Long>> bookieIdSetEntry : bookieLedgerMaps.entrySet()) {
+                Set<Long> ledgersOnBookieId = bookieIdSetEntry.getValue();
+                if (!flags.ledgerIdsOfMigratedReplica.equals(ALL)) {
```
We'd better check 'ALL' first to reduce the call of `admin.listLedgers()`

lordcheng10 replied:
admin.listLedgers(bookieIds) needs to be called whether it is ALL or not. It is currently used to build the mapping structure of ledger and bookies: toMigratedLedgerAndBookieMap.

lordcheng10 replied:
@hangc0276

admin.listLedgers(bookieIdsSet) is for building toMigratedLedgerAndBookieMap.

toMigratedLedgerAndBookieMap records the mapping of ledger and bookie for each replica to be migrated.

lordcheng10 replied:
Check 'ALL' first, does not reduce the call of admin.listLedgers()

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/zk/ZKMetadataDriverBase.java:
```diff
@@ -289,6 +297,57 @@ public void processResult(int rc, String path, Object ctx, String name) {
+
+    public CompletableFuture<Void> submitToMigrateReplicas(long ledgerId, String bookieIds) {
+        CompletableFuture<Void> createResult = new CompletableFuture<>();
+        zk.create(getLedgerForReplicasMigrationPath(ledgerId), bookieIds.getBytes(StandardCharsets.UTF_8), acls,
```
We'd better use `zk.multi()` operation to reduce the ZK RPC times.

lordcheng10 replied:
Fixed. 
PTAL,thanks!  @hangc0276

hangc0276 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicasMigrationWorker.java:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    private Replicas getReplicasToMigrate() throws InterruptedException, KeeperException {
+        List<String> ledgerIds = metadataClientDriver.listLedgersOfMigrationReplicas(replicasMigrationPath);
```
We get the whole list for each time and pick one ledger from the earliest, it will cost a lot of resources. Would you consider to tune it?

lordcheng10 replied:
OK， I will fix!

lordcheng10 replied:
Fixed.
PTAL,thanks! @hangc0276

StevenLuMT commented:
rerun failure checks

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java:
```diff
@@ -1951,6 +1953,56 @@ public int runCmd(CommandLine cmdLine) throws Exception {
+                    "The ledgerIds corresponding to the migrated replica");
+            opts.addOption("r", "readOnly", true,
+                    "Whether to set the bookie nodes of the migrated data to readOnly, the default is true");
+            opts.addOption("p", "httpPort", true, "http port,default is 8080");
```
@hangc0276
 required default is false, I didn't set this parameter, I can show that bookiesOpt is set to true:
<img width="1162" alt="image" src="https://user-images.githubusercontent.com/19296967/190057027-6fdd40bd-bd30-463b-844a-a0f132497da0.png">

lordcheng10 replied:
@hangc0276
 required default is false, I didn't set this parameter, I can show that bookiesOpt is set to true:
<img width="1162" alt="image" src="https://user-images.githubusercontent.com/19296967/190057027-6fdd40bd-bd30-463b-844a-a0f132497da0.png">

lordcheng10 replied:
Fixed. 
PTAL,thanks!  @hangc0276

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicasMigrationWorker.java:
```diff
@@ -0,0 +1,216 @@
+
+            LOG.info(String.format("Start migrate replicas(%s)!", replicas));
+            // 2. Get the fragment containing the migration bookie
+            try (LedgerHandle lh = admin.openLedgerNoRecovery(replicas.ledgerId)) {
```
We can control the speed of data migration by configuring throttle:
<img width="1159" alt="image" src="https://user-images.githubusercontent.com/19296967/190098404-9bc53d04-8204-40db-8fe1-7c553e472b8a.png">
<img width="812" alt="image" src="https://user-images.githubusercontent.com/19296967/190098793-5ca50781-89d1-47ef-a65a-fcb5ec93a09c.png">
<img width="1056" alt="image" src="https://user-images.githubusercontent.com/19296967/190098877-3fd35e04-8cdc-48be-8df2-0699aca19b7b.png">

lordcheng10 commented:
@hangc0276 @eolivelli FIxed,PTAL,thanks!

lordcheng10 commented:
rerun failure checks

lordcheng10 commented:
discuss thread：https://lists.apache.org/thread/0zlsm7z5j34j23gtlhf7rf72lfq62jyk
vote thread:  https://lists.apache.org/thread/bw4b08bhposh3gk1xzfk4sskxkjdx53b

lordcheng10 commented:
> discuss thread：https://lists.apache.org/thread/0zlsm7z5j34j23gtlhf7rf72lfq62jyk vote thread: https://lists.apache.org/thread/bw4b08bhposh3gk1xzfk4sskxkjdx53b

@eolivelli @hangc0276 @StevenLuMT @Shoothzj PTAL,thanks!

wolfstudy reviewed (approved):
LGTM +1

lordcheng10 commented:
@hangc0276 PTAL,thanks!

zymap reviewed (changes requested):
I think you need to add more details to the proposal to know how this work.
Like when will the migration does, how it does migration, and what will happen when you trigger the migration.

I take the first look at this PR. I will take another look in the next few days.

lordcheng10 commented:
> I think you need to add more details to the proposal to know how this work. Like when will the migration does, how it does migration, and what will happen when you trigger the migration.
> 
> I take the first look at this PR. I will take another look in the next few days.

OK , I will do

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,200 @@
+/**
+ * Migrate the data of the specified replica to other bookie nodes.
+ * */
+public class ReplicasMigrationCommand extends BookieCommand<ReplicasMigrationCommand.ReplicasMigrationFlags> {
```
OK , I will fix

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,200 @@
+    public boolean apply(ServerConfiguration conf, ReplicasMigrationFlags cmdFlags) {
+        try (BookKeeperAdmin admin = new BookKeeperAdmin(new ClientConfiguration(conf))) {
+            if (cmdFlags.readOnly) {
+                if (!conf.isHttpServerEnabled()) {
```
If it is not enabled, an exception will be thrown and the user will be notified.

lordcheng10 replied:
If the server is not enabled, subsequent http requests will fail, and the command will be terminated：
<img width="1378" alt="image" src="https://user-images.githubusercontent.com/19296967/194685505-66aa77af-3a13-46d7-8da7-8459a63d852c.png">

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,200 @@
+                Collection<BookieId> readOnlyBookies = admin.getReadOnlyBookies();
+
+                Set<BookieId> switchedBookies = new HashSet<>();
+                if (!switchToReadonly(admin, toSwitchBookieIdsSet, switchedBookies, true)) {
```
It is @hangc0276 's suggestion. When some bookie nodes fail to set readonly, they need to fall back to the previous state:
<img width="914" alt="image" src="https://user-images.githubusercontent.com/19296967/191399590-b8ffcb5d-651d-4fa6-b4db-8fce5c7ca1a9.png">

lordcheng10 replied:
OK, I will fixed

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,200 @@
+                            + "and the successful bookies fall back to the previous state!");
+                    Set<BookieId> fallbackSuccessfulBookies = new HashSet<>();
+                    switchedBookies.removeAll(readOnlyBookies);
+                    if (!switchToReadonly(admin, switchedBookies, fallbackSuccessfulBookies, false)) {
```
This is not a retry, but a fallback to the previous state

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataClientDriver.java:
```diff
@@ -106,6 +110,50 @@ interface SessionStateListener {
+     * */
+    default List<String> listLedgersOfMigrationReplicas(String migrationReplicasPath)
+            throws InterruptedException, KeeperException {
+        return new ArrayList<>();
```
OK， I will fixed

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/zk/ZKMetadataDriverBase.java:
```diff
@@ -289,6 +302,52 @@ public void processResult(int rc, String path, Object ctx, String name) {
     }
 
+    private String getLedgerForReplicasMigrationPath(long ledgerId) {
+        return String.format("%s/%s", replicasMigrationPath, ledgerId);
```
OK， I will fix

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/zk/ZKMetadataDriverBase.java:
```diff
@@ -239,6 +251,7 @@ protected void initialize(AbstractConfiguration<?> conf,
         }
 
         disableHealthCheckPath = ledgersRootPath + "/" + DISABLE_HEALTH_CHECK;
+        replicasMigrationPath = ledgersRootPath + "/" + MIGRATION_REPLICAS;
```
OK, I will fix

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicasMigrationWorker.java:
```diff
@@ -0,0 +1,237 @@
+/**
+ * Migrate replica data to other bookie nodes.
+ */
+public class ReplicasMigrationWorker implements Runnable {
```
OK , I will fix

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/util/BookKeeperConstants.java:
```diff
@@ -40,6 +40,7 @@ public class BookKeeperConstants {
     public static final String AVAILABLE_NODE = "available";
     public static final String COOKIE_NODE = "cookies";
     public static final String UNDER_REPLICATION_NODE = "underreplication";
+    public static final String MIGRATION_REPLICAS = "migrationReplicas";
```
OK

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java:
```diff
@@ -0,0 +1,200 @@
+                    LOG.warn(String.format("The migration task of Replicas=%s submit failed!",
+                            toMigratedLedgerAndBookieMap), e);
+                }
+                return true;
```
If no value is returned, an error will be reported.
Values are returned in many places, such as:
<img width="1172" alt="image" src="https://user-images.githubusercontent.com/19296967/191400587-28d5bf7f-5010-4eae-9d95-d39c8b12d2e5.png">

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicasMigrationWorker.java:
```diff
@@ -0,0 +1,237 @@
+                Set<LedgerFragment> fragments = getFragmentsOnMigrationBookies(lh, replicas.bookieIds);
+                if (fragments.size() < 1) {
+                    //3.The replication has been completed, delete the ledgerId directory and release lock
+                    releaseLock(replicas);
```
OK, I will fix

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented:
> In general, for the "The bookie node goes offline" use case i'd prefer if this was done via autorecovery: introduce bookie state like "draining" - read only for the client point of view, down for auditor/autorecovery purposes and use regular AR.
@dlg99 

1. The offline of the bookie node is just one of the cases solved by this data migration tool.

2.IMO, bookkeeper should have the ability to migrate data, because only data recovery is currently supported.

3. The data migration tool not only supports migrating all the data of one or more bookie nodes, but also supports migrating part of the data on one bookie node to other nodes.

4. There are two cases that we use this tool to solve:
①. The bookie node goes offline;

②. Expand the bookie node to improve the reading speed of historical data:
a. When the client consumes historical data a few days ago, it hopes to increase the reading speed of historical data by expanding the bookie node.
b. When we expand the new bookie node to the cluster, the new node can only receive the read and write of new data, and cannot improve the reading speed of historical data.
c. After the data migration tool of bookkeeper, we can migrate some historical data to the new node, let the new node provide some historical data reading, and improve the reading speed of historical data.

lordcheng10 commented:
> Overall I am ok with the approach but:
> 
> PR needs tests for the new feature
> PR needs tests that show that GC won't consider ledger as over replicated and delete the data if this kicks in at the same time as migration
> PR needs tests to show that Autorecovery can co-exist with this (e.g. ensemble size 5, migrating two bookies, plus one goes down -> AR starts rereplication, can this end up in inconsistent metadata?)
> PR needs tests for migration of an open ledger (ledger created, still open, some data written, never closed, no new writes, try to migrate it)
> possibly more tests for variety of AR/GC cases (replication worker crashes mid-migration, bookie restarts mid-migration etc)
> http API for the bookie to switch it to read-only is optional (may be disabled), that should be ok.
> the problem is that the read-only status set by the REST API is not persisted between bookie restarts IIRC (do double check me), so the bookie can restart mid-migration and become writable.
> 
> These are the areas where data loss can occur and extra attention is required.

Thanks for your suggestion! Excellent!
I will test these cases you mentioned.

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java:
```diff
@@ -1951,6 +1953,77 @@ public int runCmd(CommandLine cmdLine) throws Exception {
+            super(CMD_REPLICASMIGRATION);
+            Option bookiesOpt = new Option("b", "bookieIds", true,
+                    "Bookies corresponding to the migrated replica");
+            Option ledgersOpt = new Option("l", "ledgerIds", true,
```
Here ledgersOpt and readOnlyOpt are optional by default, and bookiesOpt is required:
<img width="1102" alt="image" src="https://user-images.githubusercontent.com/19296967/191411023-65a8bf40-31e3-4e19-b54a-df0aeabfe2fa.png">

lordcheng10 replied:
Fixed! PTAL,thanks! @zymap

lordcheng10 commented:
@hangc0276 PTAL, thanks!

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java:
```diff
@@ -1951,6 +1953,77 @@ public int runCmd(CommandLine cmdLine) throws Exception {
+            Option ledgersOpt = new Option("l", "ledgerIds", true,
+                    "The ledgerIds corresponding to the migrated replica");
+            Option readOnlyOpt = new Option("r", "readOnly", true,
+                    "Whether to set the bookie nodes of the migrated data to readOnly, the default is true");
```
OK

lordcheng10 replied:
Fixed.
PTAL,thanks! @hangc0276

lordcheng10 commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/MigrationReplicasCommand.java:
```diff
@@ -0,0 +1,189 @@
+        }
+    }
+
+    private boolean switchToReadonly(BookKeeperAdmin admin,
```
OK,I Will fix

lordcheng10 replied:
rename to "switchBookieStatus"
